### PR TITLE
Asymmetric Quantile (Pinball) Loss on Pressure Channel

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current L1 loss treats overprediction and underprediction of pressure equally. In aerodynamics, the suction peak on the upper surface (Cp << 0) is the primary driver of lift and the most error-sensitive region. A symmetric loss allows the model to "hedge" toward the mean, smoothing the suction peak.

The **pinball (quantile) loss** with tau > 0.5 penalizes underprediction more than overprediction:

```
L(y, yhat, tau) = tau * max(y - yhat, 0) + (1-tau) * max(yhat - y, 0)
```

With tau=0.65, underpredictions are penalized ~1.86× more than overpredictions. This asymmetry pushes the model toward accurately capturing the suction peak rather than smoothing it.

**This is orthogonal to the existing 2× p-channel surface weight** — it changes the *shape* of the loss, not just the *scale*. Quantile regression is a standard Kaggle technique for skewed-error targets (Koenker & Bassett, 1978).

## Instructions

Add `--pinball_pressure` flag and `--pinball_tau` hyperparameter. Apply the asymmetric loss ONLY to the pressure channel on SURFACE nodes.

### Step 1: Add arguments

```python
parser.add_argument('--pinball_pressure', action='store_true',
                    help='Use asymmetric pinball loss for pressure channel on surface nodes')
parser.add_argument('--pinball_tau', type=float, default=0.65,
                    help='Quantile parameter for pinball loss (>0.5 penalizes underprediction more)')
```

### Step 2: Implement pinball loss function

```python
def pinball_loss(pred, target, tau=0.65):
    """Asymmetric quantile (pinball) loss.
    tau > 0.5: penalize underprediction (pred < target) more.
    """
    residual = target - pred
    return torch.where(residual > 0, tau * residual, (tau - 1) * residual).mean()
```

### Step 3: Apply to pressure channel in surface loss computation

Find where the surface pressure loss is computed (look for the L1 loss computation on the pressure channel for surface nodes). Replace only the **pressure channel** component of the **surface** loss:

```python
if args.pinball_pressure:
    # For surface nodes, use pinball loss on pressure channel only
    # Keep L1 for Ux, Uy channels; keep L1 for volume nodes
    loss_p_surface = pinball_loss(pred_p_surface, target_p_surface, tau=args.pinball_tau)
else:
    loss_p_surface = F.l1_loss(pred_p_surface, target_p_surface)
```

**Critical details:**
- Apply ONLY to the pressure channel (not Ux, Uy)
- Apply ONLY to surface nodes (keep symmetric L1 for volume nodes)
- The existing 2× p-channel surface weight and pinball loss are orthogonal — both should be active simultaneously
- Do NOT change the PCGrad loss computation or any other loss component

### Step 4: Log the effective loss components

Add a W&B log for the pinball pressure loss so we can verify it's active:
```python
wandb.log({"loss_pinball_p": loss_p_surface.item()})
```

### Training commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent nezuko --seed 42 \
  --wandb_name "nezuko/pinball-pressure-s42" \
  --wandb_group "asymmetric-quantile-cp-loss" \
  --pinball_pressure --pinball_tau 0.65 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same flags, --seed 73, CUDA_VISIBLE_DEVICES=1,
#   --wandb_name "nezuko/pinball-pressure-s73"
```

### Monitoring

Watch for:
- If p_in and p_tan improve but Ux/Uy MAE degrades: tau is correct, pressure improvement is real
- If all metrics degrade: tau may be too aggressive; would try tau=0.55 next
- If metrics are flat: the model may already be well-calibrated on pressure asymmetry

## Baseline

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

**Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

**Reproduce command:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Summary: NEGATIVE — Training divergence, experiment killed early

Both seeds diverged catastrophically starting around epoch 26-27. Experiments were killed at epoch 50 (~63 min) since they were clearly unrecoverable.

### Training Trajectory

**S42 (seed 42, wandb: `hpwasesr`)**
- Stable epochs 1-25, best checkpoint at epoch 25
- Divergence onset epoch 26: val_in_dist jumps from 1.29 → 4.27
- By epoch 50: oscillating wildly (val_in_dist 3.1-14.7), train surf loss 0.18 (vs 0.009 at epoch 25)
- Never recovered

**S73 (seed 73, wandb: `75eihl1t`)**
- Stable epochs 1-26, best checkpoint at epoch 26
- Divergence onset epoch 27: val_in_dist jumps from 1.12 → 2.68
- Complete blowup by epoch 40+: val_in_dist reached **94.7** at epoch 49
- Train surf loss exploded to **2.75** (from 0.008 at epoch 26)

### Best Checkpoint Metrics (NOT comparable to baseline)

These are from early training (epoch 25-26) and represent an immature, unconverged model:

| Metric | S42 (ep25) | S73 (ep26) | Baseline |
|--------|-----------|-----------|----------|
| p_in | 1.290 | 1.119 | **11.742** |
| p_oodc | 1.214 | 1.135 | **7.643** |
| p_tan | 1.677 | 1.837 | **27.874** |
| p_re | 0.989 | 0.997 | **6.419** |

Note: These appear artificially low because they are from early training before the model has learned the hard cases. The baseline metrics are from fully converged models (~150 epochs). A fair comparison requires convergence, which pinball loss prevents.

### What happened

The pinball loss (tau=0.65) creates an **asymmetric gradient** that destabilizes training through a feedback loop with several other mechanisms:

1. **Adaptive surface weight ramp**: As surf_weight increases during training, the asymmetric pinball gradient gets amplified
2. **Hard-node mining (epoch 30+)**: The 1.5× weight on high-error surface nodes further amplifies the asymmetric signal — but instability actually begins *before* epoch 30, suggesting the core issue is pinball + surf_weight alone
3. **Tandem ramp (up to 4×)**: Tandem samples get boosted, compounding the asymmetric gradient on those samples
4. **PCGrad**: The directional bias from pinball loss creates conflicting gradient directions that PCGrad's projection can't resolve cleanly

The key evidence: S73 diverged more severely than S42 (same hyperparameters, different seed), indicating the instability is sensitive to initial conditions — a hallmark of gradient-based instability rather than a hyperparameter issue.

### Why it failed

The fundamental problem is that **tau=0.65 is too aggressive for this training setup**. The existing loss landscape already has multiple adaptive multipliers (surf_weight, tandem_ramp, hard-node mining, PCGrad). Adding an asymmetric loss on top of these creates a multiplicative instability:

- Effective asymmetry ≈ tau × surf_weight × tandem_boost × hard_node_weight
- At epoch 30+: ~0.65 × 8.0 × 4.0 × 1.5 = **31.2× effective overshoot** on underpredicted hard tandem surface nodes

This is far too aggressive for stable optimization.

### Training command

```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --seed 42 \
  --wandb_name "nezuko/pinball-pressure-s42" \
  --wandb_group "asymmetric-quantile-cp-loss" \
  --pinball_pressure --pinball_tau 0.65 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

Seed 73: same but `--seed 73`, `CUDA_VISIBLE_DEVICES=1`, wandb name `nezuko/pinball-pressure-s73`

### Peak memory: 46.1 GB (S42), 46.4 GB (S73)
### W&B runs: `hpwasesr` (S42), `75eihl1t` (S73)

### Suggested follow-ups

1. **Much lower tau (0.52-0.55)**: If asymmetry is still worth exploring, tau=0.65 is clearly too aggressive given the multiplicative interaction with existing adaptive weights. A near-symmetric tau like 0.52 would test the concept with minimal instability risk.
2. **Disable hard-node mining when using pinball**: These two mechanisms are conceptually redundant (both bias toward high-error nodes). Running pinball alone without hard-node mining would isolate the effect.
3. **Huber-pinball hybrid**: Use a smooth pinball loss (Huber-quantile) instead of the piecewise-linear version to reduce gradient discontinuity at the residual=0 boundary.
4. **Consider this a dead end**: The existing adaptive weight stack (surf_weight × tandem_ramp × hard-node × PCGrad) may already be near the stability boundary. Adding *any* asymmetric loss component may be fundamentally incompatible with this setup without major simplification.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2310 because ADVISOR: Closing — catastrophic divergence at epoch ~26-27 confirms this is fundamentally incompatible with our training setup.

**Excellent root cause analysis.** The multiplicative instability diagnosis is spot-on: tau=0.65 × surf_weight × tandem_ramp × hard_node = ~31× effective asymmetry on hard tandem surface nodes. This explodes.

Your point #4 is the right conclusion — the existing adaptive weight stack (surf_weight → tandem_ramp → hard-node mining → PCGrad) is already near the stability boundary. Adding ANY additional asymmetric loss component would require disabling parts of this stack, which would sacrifice their individual benefits. The juice isn't worth the squeeze.

Added to exhausted list: asymmetric/pinball pressure loss is incompatible with the current adaptive loss weighting stack. New experiment incoming..
